src/jit/codegenxarch.cpp

```diff
@@ -7488,115 +7514,183 @@ void CodeGen::genPutArgStk(GenTreePutArgStk* putArgStk)
+            {
+                if (m_pushStkArg)
+                {
+                    switch (fieldNode->OperGet())
+                    {
+                        case GT_LCL_VAR:
+                        case GT_LCL_FLD:
+                        case GT_LCL_VAR_ADDR:
+                            assert(varTypeIsIntegralOrI(fieldNode));
+                            inst_TT(INS_push, fieldNode, 0, 0, emitActualTypeSize(fieldNode->TypeGet()));
+                            break;
+                        case GT_CNS_INT:
+                            if (fieldNode->IsIconHandle())
+                            {
+                                inst_IV_handle(INS_push, fieldNode->gtIntCon.gtIconVal);
+                            }
+                            else
+                            {
+                                inst_IV(INS_push, fieldNode->gtIntCon.gtIconVal);
+                            }
+                            break;
+                        default:
+                            unreached();
+                    }
+                    currentOffset -= TARGET_POINTER_SIZE;
+                    genStackLevel += TARGET_POINTER_SIZE;
+                }
+                else
+                {
+                    // The stack has been adjusted and we will load the field to tmpReg and then store it on the stack.
+                    switch (fieldNode->OperGet())
+                    {
+                        case GT_LCL_VAR:
+                        case GT_LCL_FLD:
+                        case GT_LCL_VAR_ADDR:
+                            assert(varTypeIsIntegralOrI(fieldNode));
+                            inst_RV_TT(INS_mov, tmpReg, fieldNode, fieldOffset - currentOffset,
+                                       emitTypeSize(fieldNode));
+                            genStoreRegToStackArg(fieldType, tmpReg, fieldOffset - currentOffset);
+                            break;
+                        case GT_CNS_INT:
+                            inst_RV_TT(INS_mov, tmpReg, fieldNode, fieldOffset - currentOffset,
+                                       emitTypeSize(fieldNode));
+                            genStoreRegToStackArg(fieldType, tmpReg, fieldOffset - currentOffset);
+                            break;
+                        default:
+                            unreached();
+                    }
+                }
             }
             else
             {
-                assert(!m_pushStkArg);
-                genStoreRegToStackArg(fieldType, argReg, desiredOffset - currentOffset);
+                genStoreRegToStackArg(fieldType, argReg, fieldOffset - currentOffset);
+                if (m_pushStkArg)
+                {
+                    // We always push a slot-rounded size
+                    currentOffset -= genTypeSize(fieldType);
+                }
             }
 
             prevFieldOffset = fieldOffset;
         }
-
-        // Adjust the stack if necessary.
-        if (currentOffset != 0)
```

Comment:
Why was this code removed? If this situation is not thought to be possible, we should have an assert to verify that it does not occur.